tests/test_as_path_verification.c

```diff
@@ -14,108 +14,92 @@
 #include <string.h>
 #include <assert.h>
 
-static struct aspa_record *create_aspa_record(uint32_t cas, uint32_t *provider_asns, size_t provider_count)
-{
-	struct aspa_record *aspa_record = lrtr_malloc(sizeof(struct aspa_record));
-	aspa_record->customer_asn = cas;
-
-	size_t size = provider_count * sizeof(uint32_t);
-	aspa_record->provider_asns = lrtr_malloc(size);
-	memcpy(aspa_record->provider_asns, provider_asns, size);
 
-	aspa_record->provider_count = provider_count;
+#define ASNS(...) (uint32_t []) { __VA_ARGS__ }
 
-	return aspa_record;
+#define RECORD(cas, providers) (struct aspa_record) { \
+	.customer_asn = cas, \
+	.provider_count = (size_t)(sizeof(providers) / sizeof(uint32_t)), \
+	.provider_asns = sizeof(providers) == 0 ? NULL : providers \
 }
 
-static enum aspa_status aspa_store_insert_node(struct aspa_store_node **store, struct rtr_socket *rtr_socket,
-					       struct aspa_array *aspa_array)
-{
-	// Allocate new node
-	struct aspa_store_node *new = lrtr_malloc(sizeof(struct aspa_store_node));
-
-	if (new == NULL)
-		return ASPA_ERROR;
-
-	// Store socket and ASPA array
-	new->rtr_socket = rtr_socket;
-	new->aspa_array = aspa_array;
-
-	// prepend new node
-	new->next = *store; // may be NULL
-	*store = new;
-
-	return ASPA_SUCCESS;
+#define ADD_OPERATION(idx, rec) ((struct aspa_update_operation) { \
+	.index = idx, \
+	.record = rec, \
+	.type = ASPA_ADD, \
+	.is_no_op = false \
+})
+
+#define BUILD_ASPA_TABLE(...) \
+	struct aspa_table *aspa_table = lrtr_malloc(sizeof(*aspa_table)); \
+	assert(aspa_table != NULL); \
+	aspa_table_init(aspa_table, NULL); \
+	\
+	ENTER_ASPA_TABLE_NEW_SOCKET(aspa_table, __VA_ARGS__) \
+
+
+#define _CAT_(a, b) a ## b
+#define _CAT(a, b) _CAT_(a, b)
+#define ENTER_ASPA_TABLE_NEW_SOCKET(aspa_table, ...) { \
+	struct rtr_socket *_CAT(rtr_socket, __LINE__) = lrtr_malloc(sizeof(struct rtr_socket)); \
+	assert(_CAT(rtr_socket, __LINE__) != NULL); \
+	_CAT(rtr_socket, __LINE__)->aspa_table = aspa_table; \
+	\
+	struct aspa_record records[] = { __VA_ARGS__ }; \
+	size_t len = sizeof(records) / sizeof(struct aspa_record); \
+	\
+	if (len) { \
+		struct aspa_update *update = NULL; \
+		struct aspa_update_operation *operations = lrtr_malloc(len * sizeof(struct aspa_update_operation)); \
+		for (size_t i = 0; i < len; i++) \
+			operations[i] = ADD_OPERATION(i, records[i]); \
+		\
+		assert(aspa_table_update_swap_in_compute(aspa_table, _CAT(rtr_socket, __LINE__), operations, len, &update) == ASPA_SUCCESS); \
+		aspa_table_update_swap_in_apply(update); \
+		aspa_table_update_swap_in_finish(update); \
+	} \
 }
 
-static int compare_records(const void *a, const void *b)
-{
-	return (*(struct aspa_record **)a)->customer_asn - (*(struct aspa_record **)b)->customer_asn;
-}
+#define VERIFY_AS_PATH(direction, result, ...) \
```

Comment:
I'd like the third arg here not be variadic. Using the `ASNS` macro from the caller side would be more consistent with the other tests

Reply:
done